### PR TITLE
fix: Cache passed type checks for many configs

### PR DESCRIPTION
**Pull request by nayeemrmn**
Discussed on Discord a couple times.
```ts
// temp.ts
console.log("Hello, world!");

// temp.json
{
  "compilerOptions": {
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ]
  }
}
```
Before:
```
$ deno check temp.ts
Check file:///home/nayeem/projects/deno/temp.ts
$ deno check -c temp.json temp.ts
Check file:///home/nayeem/projects/deno/temp.ts
$ deno check temp.ts
Check file:///home/nayeem/projects/deno/temp.ts
$ deno check -c temp.json temp.ts
Check file:///home/nayeem/projects/deno/temp.ts
```
After:
```
$ deno check temp.ts
Check file:///home/nayeem/projects/deno/temp.ts
$ deno check -c temp.json temp.ts
Check file:///home/nayeem/projects/deno/temp.ts
$ deno check temp.ts
$ deno check -c temp.json temp.ts
```

### CODE REVIEW

**Comment by kitsonk** [comment]
I will defer to @dsherret as he has been investigating cache validation challenges as a whole.

File: cli/cache.rs [hunk 1]
```diff
@@ -14,18 +14,22 @@ use deno_graph::source::LoadFuture;
 use deno_graph::source::LoadResponse;
 use deno_graph::source::Loader;
 use deno_runtime::permissions::Permissions;
+use std::collections::HashSet;
 use std::sync::Arc;
 
 #[derive(Debug, Deserialize, Serialize)]
 pub struct EmitMetadata {
   pub version_hash: String,
+  pub type_checks_succeeded: HashSet<String>,
```
Comment (dsherret) [inline cli/cache.rs:23]: This name feels like a boolean. Any other ideas? Perhaps `type_check_hashes` with a comment saying it only contains successful hashes?

Comment (dsherret) [inline cli/cache.rs:23]: Does this handle this property not existing when deserializing?

File: cli/emit.rs [hunk 2]
```diff
@@ -126,6 +161,33 @@ impl<T: Cacher> EmitCache for T {
...
+/// type checks are performed and the second uses a subset of the same roots,
+/// it's possible that the first passes and the second fails because of missing
+/// global augmentations. But without including the roots in the config hash
+/// the second would pass the other preliminary checks and short-circuit.
+fn hash_check_config(
+  ts_config: &TsConfig,
+  mode: TypeCheckMode,
+  roots: &[(ModuleSpecifier, ModuleKind)],
+) -> String {
+  let mut input = vec![ts_config.as_bytes(), vec![mode as u8]];
```
Comment (dsherret) [inline cli/emit.rs:178]: This vector could be preallocated to prevent resizing: `let mut input = Vec::with_capacity(2 + roots.len() * 2);`

**Review by dsherret — CHANGES_REQUESTED** [review]
I think this breaks a few things, but overall I'm not sure this is worth the complexity especially now that --no-check is the default.

File: cli/emit.rs [hunk 3]
```diff
@@ -548,7 +607,7 @@ pub fn check_and_maybe_emit(
         let mut emit_data_item = emit_data_items
           .entry(specifier.clone())
           .or_insert_with(|| SpecifierEmitData {
-            version_hash: get_version(source_bytes, &config_bytes),
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/emit.rs:551]: It makes caching a lot more complicated and I'm still not sure it's worth it. It seems much easier and less error prone to just invalidate all the caches whenever the config changes. How often do people have multiple configs or no config and then a config when type checking? Is this actually something users need?

[the remark this answers is not in the document]
Comment (nayeemrmn) [reply cli/emit.rs:551]: It happens with workers also

**Comment by nayeemrmn** [comment]
@dsherret I remembered why I wanted to do this, last time it came up was discussing #14632. We could fix shortly fix that just by including the check scope in the config digest. But looking at the cache validation code I strongly think we need better separation between things that bust the emit cache, when a type check needs to be reperformed and tracking if the source has actually changed. As described in https://github.com/denoland/deno/pull/15090#discussion_r915235504. One of the reasons is that invalid emits from changed sources need to be checked for each module in the graph uniformly, whereas checked state can only correctly be stored at the roots. Another is that some files aren't emitted at all but changes need to be tracked only for type checking.

There have always been odd cache validation bugs and I attribute it to the current structure which conflates invalid emits and yet-to-be-performed type checks and tries to store it all in one version hash. But again I can make a better case and implementation for this _after_ checking and emitting are properly decoupled with swc. Closing for now.

**Comment by dsherret** [comment]
@nayeemrmn yeah, I think decoupling the two would be really good. I was planning on starting work tomorrow on #13302. Maybe at the same time I will look a bit at decoupling it.

**Comment by dsherret** [comment]
I looked into this today and the type checking cache seems like it won't be invalidated in more scenarios than mentioned in the original issue. I'm updating the code to cache a hash of all the inputs per successful type check which is only cleared when changing CLI versions, so it will solve this issue too.

**Pull request closed without merging**